lib/app/server.ts

```diff
@@ -0,0 +1,437 @@
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import fs from 'node:fs/promises';
+import path from 'node:path';
+import process from 'node:process';
+import url from 'node:url';
+
+import * as Sentry from '@sentry/node';
+import compression from 'compression';
+import express from 'express';
+import type {NextFunction, Request, Response} from 'express';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import morgan from 'morgan';
+import PromClient from 'prom-client';
+import responseTime from 'response-time';
+import sanitize from 'sanitize-filename';
+import sFavicon from 'serve-favicon';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import systemdSocket from 'systemd-socket';
+import _ from 'underscore';
+import urljoin from 'url-join';
+
+import {ElementType} from '../../shared/common-utils.js';
+import type {AppArguments} from '../app.interfaces.js';
+import {GoldenLayoutRootStruct} from '../clientstate-normalizer.js';
+import * as normalizer from '../clientstate-normalizer.js';
+import type {ShortLinkMetaData} from '../handlers/handler.interfaces.js';
+import {cached, csp} from '../handlers/middleware.js';
+import {logger, makeLogStream} from '../logger.js';
+import {ShortLinkResolver} from '../shortener/google.js';
+import * as utils from '../utils.js';
+import type {
+    PugRequireHandler,
+    RenderConfig,
+    ServerDependencies,
+    ServerOptions,
+    WebServerResult,
+} from './server.interfaces.js';
+import {getFaviconFilename, isDevMode} from './utils.js';
+
+/**
+ * Detects if the request is from a mobile viewer
+ */
+export function isMobileViewer(req: Request): boolean {
+    return req.header('CloudFront-Is-Mobile-Viewer') === 'true';
+}
+
+/**
+ * Configure a web server and its routes
+ */
+export async function setupWebServer(
+    appArgs: AppArguments,
+    options: ServerOptions,
+    dependencies: ServerDependencies,
+): Promise<WebServerResult> {
+    const {ceProps, sponsorConfig, clientOptionsHandler} = dependencies;
+    const {
+        staticPath,
+        staticMaxAgeSecs,
+        staticUrl,
+        staticRoot,
+        httpRoot,
+        sentrySlowRequestMs,
+        distPath,
+        extraBodyClass,
+        maxUploadSize,
+    } = options;
+
+    // Create the Express application and router
+    const webServer = express();
+    const router = express.Router();
+
+    // Initialize services
+    const googleShortUrlResolver = new ShortLinkResolver();
+    let pugRequireHandler: PugRequireHandler = () => {
+        logger.error('pug require handler not configured');
+        return '';
+    };
+
+    /**
+     * Renders configuration for templates
+     */
+    function renderConfig(extra: Record<string, any>, urlOptions?: Record<string, any>): RenderConfig {
```

Comment:
Maybe it makes sense to make this a top-level function? createRenderConfig() that optionally takes the variables capture by closure here

Reply:
It's a bit of a change to do this: the function is passed to lots of places around the codebase that assuems it can "just" call this function. It has to fit the API of `(extra, url?) => Config` for those places.  I'd prefer to do that in a separate change as it affects more code around the codebase.